compiler/src/iree/compiler/Dialect/HAL/Transforms/ResolveTopologyQueries.cpp

```diff
@@ -0,0 +1,228 @@
+// The bits set here represent the superset of required and allowed bits and
+// are useful for providing buffers back to users via the ABI that may need to
+// be used for more than just what the internal program requires.
+LogicalResult
+deriveAllowedResourceBufferBits(Location loc, IREE::HAL::Lifetime lifetime,
+                                IREE::HAL::MemoryTypeBitfield &memoryTypes,
+                                IREE::HAL::BufferUsageBitfield &bufferUsage) {
+  memoryTypes = IREE::HAL::MemoryTypeBitfield::None;
+  bufferUsage = IREE::HAL::BufferUsageBitfield::None;
+  if (failed(deriveRequiredResourceBufferBits(loc, lifetime, memoryTypes,
+                                              bufferUsage))) {
+    return failure();
+  }
+  switch (lifetime) {
+  default:
+    break;
+  case IREE::HAL::Lifetime::External:
+    if (clExternalResourcesMappable) {
+      // #yolo; these come from/go to outside the program.
+      // Today we assume they are device-local|host-visible just for
+      // practical purposes but that does not have to be true. We really
+      // want this to be something we analyze and handle on the edges
+      // (transferring devices/etc if needed).
+      memoryTypes = memoryTypes | IREE::HAL::MemoryTypeBitfield::DeviceLocal |
+                    IREE::HAL::MemoryTypeBitfield::HostVisible;
+      // NOTE: we may not map it but users may after they get them back.
+      // Another reason we should annotate this - having a buffer be
+      // mappable is potentially expensive (may get a 2nd copy in memory!).
+      bufferUsage = bufferUsage | IREE::HAL::BufferUsageBitfield::Mapping;
+    } else {
+      memoryTypes = memoryTypes | IREE::HAL::MemoryTypeBitfield::DeviceLocal;
+    }
+    break;
+  }
+  return success();
+}
+
+// Checks if all affinities of the optimal attribute refer to the same device
+// ID.
+static bool allReferToSameDevice(DeviceOptimalAttr optimalAttr,
+                                 DeviceAnalysis &deviceAnalysis,
+                                 Operation *fromOp) {
+  SetVector<IREE::HAL::DeviceTargetAttr> allPossibleTargets;
+  for (auto affinity : optimalAttr.getAffinities()) {
+    deviceAnalysis.gatherDeviceAffinityTargets(affinity, fromOp,
+                                               allPossibleTargets);
+  }
+  if (allPossibleTargets.empty()) {
+    return false;
+  }
+  StringAttr firstDeviceId = allPossibleTargets.front().getDeviceID();
+  for (IREE::HAL::DeviceTargetAttr target : allPossibleTargets) {
+    if (firstDeviceId != target.getDeviceID()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Checks if the given device has transparent access to all other devices.
+static bool hasTransparentAccessToAll(IREE::HAL::DeviceTopologyAttr topology,
+                                      IREE::Stream::AffinityAttr source,
+                                      DeviceOptimalAttr optimalAttr) {
+  LLVM_DEBUG(llvm::dbgs() << "[resolve_topology_queries] Checking if " << source
```

Comment:
common convention in LLVM/MLIR is to use pass name substrings (so searching "resolve-topology-queries" if reading a debug output would point back to the pass, or if looking for the outputs from a pass you can type its identifier - differing makes that harder)

```suggestion
  LLVM_DEBUG(llvm::dbgs() << "[resolve-topology-queries] Checking if " << source
```

(here and elsewhere)